src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SNI/SNICommon.cs

```diff
@@ -397,5 +355,35 @@ internal static uint ReportSNIError(SNIError error)
             SNILoadHandle.SingletonInstance.LastError = error;
             return TdsEnums.SNI_ERROR;
         }
+
+        internal static unsafe bool ByteArrayCompare(byte[] original, byte[] variant)
```

Comment:
I've just re-run an older benchmark, and I don't think you need to include this. I received these results from an 8K byte array.

| Method                             | Comparison | Mean           | Error       | StdDev      |
|----------------------------------- |----------- |---------------:|------------:|------------:|
| ByteByByteComparison             | MatchingArrays | 11,499.0741 ns | 221.0199 ns | 195.9284 ns |
| ByteByByteComparison             | MismatchingArrays |  6,755.5585 ns |  43.1916 ns |  40.4015 ns |
| SpanSequenceEqualComparison                   | MatchingArrays |    232.1920 ns |   1.7383 ns |   1.4515 ns |
| SpanSequenceEqualComparison                   | MismatchingArrays |    232.1477 ns |   2.6315 ns |   2.1974 ns |
| PR2439Comparison | MatchingArrays |    623.6450 ns |  41.8087 ns | 123.2739 ns |
| PR2439Comparison | MismatchingArrays |    576.4529 ns |   4.0778 ns |   3.8143 ns |

`original.AsSpan().SequenceEqual(variant.AsSpan())` is faster, doesn't need unsafe code and removes the LINQ reference which is dragged in by `original.SequenceEqual(variant)`.

Incidentally, it's faster than comparing long-by-long, int-by-int, short-by-short and byte-by-byte because .NET can start to use intrinsics like AVX2 and AVX512 to compare 32 or 64 bytes at a time.

Reply:
Unless this certificate validation happens extremely frequently is is incredibly slow i'm not sure the performance difference of the method used makes a difference. As long as it's the correct operation you could just use a for loop to compare the bytes if you wanted and it would be sufficient.

If we're already using the package that provides MemoryMarshal for netstandard and netfx then i'd say use MemoryMashal, otherwise just write a loop.

Reply:
AsSpan().SequenceEqual is now in use.